From 0534fac43320812499d9a7890c23b0a04be40d78 Mon Sep 17 00:00:00 2001
From: Iain Sandoe <iain@sandoe.co.uk>
Date: Thu, 27 Aug 2020 19:39:10 +0100
Subject: [PATCH] Darwin, Arm64 : Make code label references use PAGE relocs

We need to be able to refer to Lxxxx labels in libgnat.

(cherry picked from commit 80a208d8f5d105bcca03de83c20eb020a0e6fcdf)
---
 gcc/config/aarch64/aarch64.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git gcc/config/aarch64/aarch64.c gcc/config/aarch64/aarch64.c
index 0cd8026051f..44d59930401 100644
--- gcc/config/aarch64/aarch64.c
+++ gcc/config/aarch64/aarch64.c
@@ -10274,6 +10274,7 @@ output_macho_postfix_expr (FILE *file, rtx x, const char *postfix)
     case CODE_LABEL:
       ASM_GENERATE_INTERNAL_LABEL (buf, "L", CODE_LABEL_NUMBER (x));
       assemble_name (file, buf);
+      fprintf (file, "@%s", postfix);
       break;
 
     case CONST_INT:
@@ -16483,7 +16484,9 @@ aarch64_classify_symbol (rtx x, HOST_WIDE_INT offset)
 	case AARCH64_CMODEL_SMALL_SPIC:
 	case AARCH64_CMODEL_SMALL_PIC:
 	case AARCH64_CMODEL_SMALL:
-	  return SYMBOL_SMALL_ABSOLUTE;
+	  return TARGET_MACHO
+		 ? SYMBOL_MO_SMALL_PCR
+		 : SYMBOL_SMALL_ABSOLUTE;
 
 	default:
 	  gcc_unreachable ();
-- 
2.40.1

